apps/studio/src/lib/db/clients/clickhouse/ClickHouseCursor.ts

```diff
@@ -0,0 +1,100 @@
+import { BeeCursor } from "../../models";
+import rawlog from "electron-log";
+import type { ClickHouseClient, Row, StreamReadable } from "@clickhouse/client";
+import { uuidv4 } from "@/lib/uuid";
+import { waitFor } from "../base/wait";
+
+interface CursorOptions {
+  query: string;
+  params: Record<string, any>;
+  client: ClickHouseClient;
+  chunkSize: number;
+}
+
+const log = rawlog.scope("clickhouse/cursor");
+
+export class ClickHouseCursor extends BeeCursor {
+  private readonly options: CursorOptions;
+  private client: ClickHouseClient;
+  private queryId: string;
+  private stream: StreamReadable<Row<unknown, "JSONCompactEachRow">[]>;
+  private rowBuffer: any[][] = [];
+  private end = false;
+  private error?: Error;
+  private bufferReady = false;
+
+  constructor(options: CursorOptions) {
+    super(options.chunkSize);
+    this.options = options;
+    this.client = options.client;
+  }
+
+  async start(): Promise<void> {
+    log.info("Starting cursor");
+
+    this.queryId = uuidv4();
+
+    const result = await this.client.query({
+      query: this.options.query,
+      query_params: this.options.params,
+      query_id: this.queryId,
+      format: "JSONCompactEachRow",
+    });
+
+    this.stream = result.stream();
+
+    this.stream.on("data", this.handleRows.bind(this));
+    this.stream.on("end", this.handleEnd.bind(this));
+    this.stream.on("error", this.handleError.bind(this));
+  }
+
+  private handleRows(rows: Row[]) {
```

Comment:
@rathboma This is probably the closest implementation of the cursor. I used the exact codes from `MysqlCursor`. But the only different thing is `ClickHouse` stream returns the data in "chunks" which can contain a size of rows that may vary instead of "a single row". Unfortunately, I couldn't find a way to control it.

I tried [max_block_size](https://clickhouse.com/docs/en/operations/settings/settings#setting-max_block_size) and [limit](https://clickhouse.com/docs/en/operations/settings/settings#limit) to set the chunk size but those don't work like how we want.

I can fake the `chunkSize` by making a new array or something else but are we OK to ignore `chunkSize` actually? Maybe it's possible to make a flag that says "this driver doesn't support chunkSize stream"?